File: app/src/main/java/com/hsilighting/pagify/ui/library/LibraryScreen.kt
package com.hsilighting.pagify.ui.library

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Description
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.platform.LocalSoftwareKeyboardController
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.hsilighting.pagify.R
import com.hsilighting.pagify.core.RecentDocument
import com.hsilighting.pagify.core.recentSubtitle
import com.hsilighting.pagify.core.searchRecents
import com.hsilighting.pagify.ui.components.combinedClickableCompat
import com.hsilighting.pagify.ui.components.longPressHint

/**
 * What the app opens on: the documents you have read before.
 *
 * The reader used to be the front door, which meant every launch began with an
 * empty grey screen and a file picker — the app asking a question it already knew
 * the answer to. Almost every session is the document from last time, or the one
 * before it.
 *
 * Only documents that actually opened are here. A file that failed, or that asked
 * for a password and never got one, is not something to offer again as though it
 * worked.
 */
@Composable
fun LibraryScreen(
    documents: List<RecentDocument>,
    onOpen: (RecentDocument) -> Unit,
    onForget: (RecentDocument) -> Unit,
    onPickDocument: () -> Unit,
    modifier: Modifier = Modifier,
) {
    var query by remember { mutableStateOf("") }
    val search = remember { FocusRequester() }
    val keyboard = LocalSoftwareKeyboardController.current
    val shown = remember(documents, query) { searchRecents(documents, query) }

    Box(modifier.fillMaxSize()) {
        Column(Modifier.fillMaxSize()) {
            Header(onSearch = { runCatching { search.requestFocus() } })

            Text(
                "Document Library",
                style = MaterialTheme.typography.headlineSmall,
                fontWeight = FontWeight.Bold,
                modifier = Modifier.padding(start = 20.dp, end = 20.dp, bottom = 12.dp),
            )

            OutlinedTextField(
                value = query,
                onValueChange = { query = it },
                singleLine = true,
                shape = RoundedCornerShape(14.dp),
                placeholder = { Text("Search files…") },
                leadingIcon = { Icon(Icons.Filled.Search, contentDescription = null) },
                trailingIcon = {
                    if (query.isNotEmpty()) {
                        IconButton(onClick = { query = "" }) {
                            Icon(Icons.Filled.Close, contentDescription = "Clear the search")
                        }
                    }
                },
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Search),
                keyboardActions = KeyboardActions(onSearch = { keyboard?.hide() }),
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 20.dp)
                    .focusRequester(search),
            )

            Spacer(Modifier.height(12.dp))

            when {
                documents.isEmpty() -> EmptyLibrary(onPickDocument)

                shown.isEmpty() -> NothingMatched(query)

                else -> LazyColumn(
                    // Room at the bottom for the button that floats over it.
                    contentPadding = PaddingValues(start = 16.dp, end = 16.dp, bottom = 96.dp),
                    verticalArrangement = Arrangement.spacedBy(10.dp),
                ) {
                    items(shown, key = { it.uri }) { document ->
                        DocumentRow(
                            document = document,
                            onOpen = { onOpen(document) },
                            onForget = { onForget(document) },
                        )
                    }
                }
            }
        }

        FloatingActionButton(
            onClick = onPickDocument,
            modifier = Modifier
                .align(Alignment.BottomEnd)
                .padding(20.dp),
        ) {
            Icon(Icons.Filled.Add, contentDescription = "Open a PDF")
        }
    }
}

/**
 * The app's name and mark.
 *
 * The launcher icon rather than a second drawing of it: the thing someone tapped
 * to get here is the thing that should greet them, and two versions of a logo
 * drift apart the moment one of them is edited.
 */
@Composable
private fun Header(onSearch: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 20.dp, vertical = 16.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        // The adaptive icon's two layers, composed by hand. `painterResource`
        // cannot load an adaptive icon — it throws "Only VectorDrawables and
        // rasterized asset types are supported" and takes the app down with it.
        //
        // The foreground is drawn at the full size of the circle rather than
        // scaled up the way a launcher mask does: this artwork is already drawn
        // to the edge of its canvas, so enlarging it clips the P and the corner
        // of the page.
        Box(
            modifier = Modifier
                .size(LOGO_SIZE)
                .clip(CircleShape)
                .background(colorResource(R.color.ic_launcher_background)),
            contentAlignment = Alignment.Center,
        ) {
            Image(
                painter = painterResource(R.drawable.ic_launcher_foreground),
                contentDescription = null,
                modifier = Modifier.size(LOGO_SIZE),
            )
        }
        Text(
            "Pagify",
            style = MaterialTheme.typography.titleLarge,
            fontWeight = FontWeight.Bold,
            color = MaterialTheme.colorScheme.primary,
            modifier = Modifier.weight(1f),
        )
        IconButton(onClick = onSearch) {
            Icon(Icons.Filled.Search, contentDescription = "Search the library")
        }
    }
}

/**
 * One document: what it is called, and enough about it to recognise it.
 *
 * Long press offers to forget it. A row can outlive the file it points at — moved,
 * deleted, or a grant a reboot took away — and without a way to remove it the
 * library slowly fills with rows that only ever fail.
 */
@Composable
private fun DocumentRow(
    document: RecentDocument,
    onOpen: () -> Unit,
    onForget: () -> Unit,
) {
    var menuOpen by remember { mutableStateOf(false) }

    Surface(
        shape = RoundedCornerShape(16.dp),
        color = MaterialTheme.colorScheme.surface,
        tonalElevation = 1.dp,
        shadowElevation = 1.dp,
        modifier = Modifier
            .fillMaxWidth()
            .combinedClickableCompat(onClick = onOpen, onLongClick = { menuOpen = true }),
    ) {
        Row(
            modifier = Modifier.padding(14.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(
                modifier = Modifier
                    .size(46.dp)
                    .clip(RoundedCornerShape(10.dp))
                    .background(MaterialTheme.colorScheme.surfaceVariant)
                    // The row answers a long press with a menu, and the tile is
                    // the only part of it shaped like an icon.
                    .longPressHint(
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                        inset = 4.dp,
                    ),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    Icons.Filled.Description,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            Column(
                modifier = Modifier
                    .weight(1f)
                    .padding(start = 14.dp),
            ) {
                Text(
                    document.name,
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                val subtitle = recentSubtitle(document)
                if (subtitle.isNotEmpty()) {
                    Text(
                        subtitle,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
            }

            DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                DropdownMenuItem(
                    text = { Text("Remove from library") },
                    onClick = {
                        menuOpen = false
                        onForget()
                    },
                )
            }
        }
    }
}

@Composable
private fun EmptyLibrary(onPickDocument: () -> Unit) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Text(
            "Nothing here yet",
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.SemiBold,
        )
        Spacer(Modifier.height(6.dp))
        Text(
            "Documents you open show up here, newest first.",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Spacer(Modifier.height(16.dp))
        TextButton(onClick = onPickDocument) { Text("Open a PDF") }
    }
}

@Composable
private fun NothingMatched(query: String) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Text(
            "No document called “$query”",
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.SemiBold,
        )
        Spacer(Modifier.height(6.dp))
        Text(
            "The library only holds documents you have opened before.",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

/** The mark in the header, matching the size of a toolbar icon and its label. */
private val LOGO_SIZE = 34.dp

File: ios/Sources/LibraryScreen.swift
import SwiftUI

/// What the app opens on: the documents you have read before.
///
/// The reader used to be the front door, which meant every launch began with an
/// empty grey screen and a file picker — the app asking a question it already
/// knew the answer to. Almost every session is the document from last time, or
/// the one before it.
///
/// Only documents that actually opened are here. A file that failed, or that
/// asked for a password and never got one, is not something to offer again as
/// though it worked.
struct LibraryScreen: View {
    let documents: [RecentDocument]
    let onOpen: (RecentDocument) -> Void
    let onForget: (RecentDocument) -> Void
    /// Show the chooser. Both the button that floats over the list and the one
    /// on the empty screen ask this same question — "a document you have, or one
    /// that does not exist yet?" — so neither of them reaches the file picker
    /// directly.
    let onPickDocument: () -> Void

    @Environment(\.colorScheme) private var scheme
    @State private var query = ""
    @FocusState private var searchFocused: Bool

    private var shown: [RecentDocument] {
        searchRecents(documents, query: query)
    }

    var body: some View {
        ZStack(alignment: .bottomTrailing) {
            VStack(spacing: 0) {
                header

                Text("Document Library")
                    .font(.title2.weight(.bold))
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.horizontal, 20)
                    .padding(.bottom, 12)

                searchField

                // A fixed gap. `Spacer(minLength:)` is flexible, and in a column
                // that fills the screen it grows to swallow whatever is left —
                // pushing the list down past the fold on a tall phone.
                Spacer().frame(height: 12)

                if documents.isEmpty {
                    EmptyLibrary(onPickDocument: onPickDocument)
                } else if shown.isEmpty {
                    NothingMatched(query: query)
                } else {
                    list
                }
            }

            Button(action: onPickDocument) {
                Image(systemName: "plus")
                    .font(.system(size: 22, weight: .semibold))
                    .frame(width: 56, height: 56)
                    .background(PagifyColor.primary(scheme),
                                in: RoundedRectangle(cornerRadius: 17))
                    .foregroundStyle(PagifyColor.onPrimary(scheme))
                    .shadow(color: .black.opacity(0.2), radius: 5, y: 2)
            }
            .padding(20)
            .accessibilityLabel("Add a document")
        }
        .background(PagifyColor.background(scheme))
    }

    private var list: some View {
        ScrollView {
            LazyVStack(spacing: 10) {
                ForEach(shown) { document in
                    Button {
                        onOpen(document)
                    } label: {
                        DocumentRow(document: document)
                    }
                    .buttonStyle(.plain)
                    // Long press is the single path to forgetting a row, the way
                    // it is on Android. A swipe action would be a second one,
                    // and a destructive action with two doors is a destructive
                    // action people find by accident.
                    .contextMenu {
                        Button(role: .destructive) {
                            onForget(document)
                        } label: {
                            Label("Remove from library", systemImage: "trash")
                        }
                    }
                }
            }
            .padding(.horizontal, 16)
            // Room at the bottom for the button that floats over it.
            .padding(.bottom, 96)
        }
    }

    /// The app's name and mark.
    ///
    /// The launcher artwork rather than a second drawing of it: the thing
    /// someone tapped to get here is the thing that should greet them, and two
    /// versions of a logo drift apart the moment one of them is edited.
    private var header: some View {
        HStack(spacing: 12) {
            ZStack {
                Circle().fill(Color(hex: 0x3B00E6))
                Image("LibraryMark")
                    .resizable()
                    .frame(width: logoSize, height: logoSize)
            }
            .frame(width: logoSize, height: logoSize)

            Text("Pagify")
                .font(.title3.weight(.bold))
                .foregroundStyle(PagifyColor.primary(scheme))

            Spacer()

            Button { searchFocused = true } label: {
                Image(systemName: "magnifyingglass")
            }
            .accessibilityLabel("Search the library")
        }
        .padding(.horizontal, 20)
        .padding(.vertical, 16)
    }

    private var searchField: some View {
        HStack(spacing: 8) {
            Image(systemName: "magnifyingglass")
                .foregroundStyle(PagifyColor.onSurfaceVariant(scheme))
            TextField("Search files\u{2026}", text: $query)
                .focused($searchFocused)
                .submitLabel(.search)
                // Searching is what the return key means here; there is nothing
                // to submit, so it only has the keyboard to put away.
                .onSubmit { searchFocused = false }
            if !query.isEmpty {
                Button { query = "" } label: {
                    Image(systemName: "xmark.circle.fill")
                        .foregroundStyle(PagifyColor.onSurfaceVariant(scheme))
                }
                .accessibilityLabel("Clear the search")
            }
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 11)
        .background(PagifyColor.surface(scheme), in: RoundedRectangle(cornerRadius: 14))
        .overlay(RoundedRectangle(cornerRadius: 14)
            .strokeBorder(Color.secondary.opacity(0.25)))
        .padding(.horizontal, 20)
    }
}

/// One document: what it is called, and enough about it to recognise it.
///
/// Long press offers to forget it. A row can outlive the file it points at —
/// moved, deleted, or a bookmark that no longer resolves — and without a way to
/// remove it the library slowly fills with rows that only ever fail.
private struct DocumentRow: View {
    @Environment(\.colorScheme) private var scheme
    let document: RecentDocument

    var body: some View {
        HStack(spacing: 14) {
            Image(systemName: "doc.text")
                .font(.system(size: 20))
                .foregroundStyle(PagifyColor.onSurfaceVariant(scheme))
                .frame(width: 46, height: 46)
                .background(PagifyColor.surfaceVariant(scheme),
                            in: RoundedRectangle(cornerRadius: 10))
                // The row answers a long press with a menu, and the tile is the
                // only part of it shaped like an icon.
                .overlay(LongPressHint().fill(PagifyColor.onSurfaceVariant(scheme)))

            VStack(alignment: .leading, spacing: 2) {
                Text(document.name)
                    .font(.subheadline.weight(.semibold))
                    .lineLimit(1)
                    .truncationMode(.tail)

                let subtitle = recentSubtitle(document)
                if !subtitle.isEmpty {
                    Text(subtitle)
                        .font(.caption)
                        .foregroundStyle(PagifyColor.onSurfaceVariant(scheme))
                        .lineLimit(1)
                        .truncationMode(.tail)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(14)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(PagifyColor.surface(scheme), in: RoundedRectangle(cornerRadius: 16))
        .contentShape(Rectangle())
    }
}

/// The mark that says "there is more under this one".
///
/// A small wedge in the bottom-right corner, the way a menu that opens on a long
/// press has been signposted since long before either platform. It exists
/// because a long press is invisible: a control that only answers to one is a
/// control most people never find.
///
/// Drawn rather than a glyph: it is three points, it has to sit exactly in the
/// corner of whatever it marks, and an SF Symbol would bring its own padding and
/// its own baseline to argue with.
private struct LongPressHint: Shape {
    var size: CGFloat = 6
    var inset: CGFloat = 4

    func path(in rect: CGRect) -> Path {
        let right = rect.maxX - inset
        let bottom = rect.maxY - inset

        // A right triangle filling the corner: across, down, and back to the
        // point. The hypotenuse faces up and left, which is what makes it read
        // as an arrow aimed at the corner rather than as a stray dot.
        var path = Path()
        path.move(to: CGPoint(x: right, y: bottom - size))
        path.addLine(to: CGPoint(x: right, y: bottom))
        path.addLine(to: CGPoint(x: right - size, y: bottom))
        path.closeSubpath()
        return path
    }
}

private struct EmptyLibrary: View {
    @Environment(\.colorScheme) private var scheme
    let onPickDocument: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            Text("Nothing here yet")
                .font(.headline)
            Spacer().frame(height: 6)
            Text("Documents you open show up here, newest first.")
                .font(.subheadline)
                .foregroundStyle(PagifyColor.onSurfaceVariant(scheme))
                .multilineTextAlignment(.center)
            Spacer().frame(height: 16)
            Button("Open a PDF", action: onPickDocument)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .padding(32)
    }
}

/// What a search that found nothing says.
///
/// The query is quoted back untrimmed and unaltered, because the surprise is
/// usually in what was typed — a stray space, a paste that brought a newline —
/// and a tidied-up echo hides exactly the thing that would explain the result.
/// There is no button: nothing here can be opened, and offering the file picker
/// would answer a question nobody asked.
private struct NothingMatched: View {
    @Environment(\.colorScheme) private var scheme
    let query: String

    var body: some View {
        VStack(spacing: 0) {
            Text("No document called \u{201C}\(query)\u{201D}")
                .font(.headline)
                .multilineTextAlignment(.center)
            Spacer().frame(height: 6)
            Text("The library only holds documents you have opened before.")
                .font(.subheadline)
                .foregroundStyle(PagifyColor.onSurfaceVariant(scheme))
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .padding(32)
    }
}

/// The mark in the header, matching the size of a toolbar icon and its label.
private let logoSize: CGFloat = 34
